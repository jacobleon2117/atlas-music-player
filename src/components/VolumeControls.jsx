import { Volume2 } from "lucide-react";

export default function VolumeControl() {
  return (
    <div className="flex items-center gap-4" aria-label="Volume controls">
      <Volume2
        className="h-6 w-6 text-black"
        fill="currentColor"
        aria-label="Volume icon"
      />
      <div
        className="bg-slider-bg relative h-2 w-[374px] rounded-lg"
        aria-label="Volume slider track"
      >
        <div
          className="bg-custom-slate absolute h-full w-1/2 rounded-l-lg"
          aria-label="Volume slider progress"
        />
        <div
          className="bg-custom-slate absolute left-1/2 top-1/2 h-4 w-4 -translate-x-1/2 -translate-y-1/2 rounded-full"
          aria-label="Volume slider handle"
        />
      </div>
    </div>
  );
}
