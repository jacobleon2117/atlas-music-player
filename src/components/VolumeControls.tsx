import { useState, useRef, useEffect } from "react";
import { Volume2, VolumeX } from "lucide-react";

interface VolumeControlsProps {
  volume: number;
  setVolume: (volume: number) => void;
}

export default function VolumeControls({ volume, setVolume }: VolumeControlsProps) {
  const [isMuted, setIsMuted] = useState(false);
  const [isDragging, setIsDragging] = useState(false);
  const sliderRef = useRef<HTMLDivElement>(null);
  const handleRef = useRef<HTMLDivElement>(null);

  const calculateVolume = (clientX: number): number => {
    if (!sliderRef.current || !handleRef.current) return 50;
    
    const rect = sliderRef.current.getBoundingClientRect();
    const handleWidth = handleRef.current.offsetWidth;
    const adjustedLeft = rect.left + handleWidth / 2;
    const adjustedWidth = rect.width - handleWidth;
    const x = clientX - adjustedLeft;
    return Math.min(Math.max((x / adjustedWidth) * 100, 0), 100);
  };

  const handleVolumeChange = (clientX: number) => {
    const newVolume = calculateVolume(clientX);
    setVolume(newVolume);
    if (isMuted) setIsMuted(false);
  };

  const handleMouseDown = (e: React.MouseEvent) => {
    setIsDragging(true);
    handleVolumeChange(e.clientX);
  };

  const handleMouseMove = (e: MouseEvent) => {
    if (!isDragging) return;
    handleVolumeChange(e.clientX);
  };

  const handleMouseUp = () => {
    setIsDragging(false);
  };

  const handleTouchMove = (e: TouchEvent) => {
    if (!isDragging) return;
    e.preventDefault();
    const touch = e.touches[0];
    handleVolumeChange(touch.clientX);
  };

  const handleTouchStart = (e: React.TouchEvent) => {
    setIsDragging(true);
    const touch = e.touches[0];
    handleVolumeChange(touch.clientX);
  };

  useEffect(() => {
    if (isDragging) {
      document.addEventListener("mousemove", handleMouseMove);
      document.addEventListener("mouseup", handleMouseUp);
      document.addEventListener("touchmove", handleTouchMove, {
        passive: false,
      });
      document.addEventListener("touchend", handleMouseUp);
    }

    return () => {
      document.removeEventListener("mousemove", handleMouseMove);
      document.removeEventListener("mouseup", handleMouseUp);
      document.removeEventListener("touchmove", handleTouchMove);
      document.removeEventListener("touchend", handleMouseUp);
    };
  }, [isDragging]);

  return (
    <div className="flex items-center gap-4" aria-label="Volume controls">
      <button
        onClick={() => setIsMuted(!isMuted)}
        className="text-primary-dark transition-colors hover:text-primary-sage dark:text-primary-light dark:hover:text-primary-sage"
        aria-label={isMuted ? "Unmute" : "Mute"}
      >
        {isMuted ? (
          <VolumeX className="h-6 w-6" />
        ) : (
          <Volume2 className="h-6 w-6" />
        )}
      </button>
      <div
        ref={sliderRef}
        onMouseDown={handleMouseDown}
        onTouchStart={handleTouchStart}
        className="relative h-2 w-[374px] cursor-pointer rounded-lg bg-primary-light dark:bg-primary-sage/30"
        role="slider"
        aria-label="Volume slider"
        aria-valuemin={0}
        aria-valuemax={100}
        aria-valuenow={isMuted ? 0 : volume}
      >
        <div
          className="absolute h-full rounded-l-lg bg-primary-sage transition-[width] duration-75 dark:bg-primary-light"
          style={{ width: `${isMuted ? 0 : volume}%` }}
        />
        <div
          ref={handleRef}
          className="absolute top-1/2 h-4 w-4 -translate-y-1/2 rounded-full bg-primary-sage shadow-md transition-transform duration-75 hover:scale-110 dark:bg-primary-light"
          style={{
            left: `${isMuted ? 0 : volume}%`,
            transform: "translate(-50%, -50%)",
          }}
        />
      </div>
    </div>
  );
}