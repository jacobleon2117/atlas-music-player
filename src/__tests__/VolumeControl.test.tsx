import { describe, it, expect } from 'vitest';
import { render } from '@testing-library/react';
import VolumeControls from '../components/VolumeControls';

describe('VolumeControls', () => {
  it('renders with default volume', () => {
    const { container } = render(
      <VolumeControls
        volume={50}
        setVolume={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with zero volume', () => {
    const { container } = render(
      <VolumeControls
        volume={0}
        setVolume={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('renders with max volume', () => {
    const { container } = render(
      <VolumeControls
        volume={100}
        setVolume={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });
});