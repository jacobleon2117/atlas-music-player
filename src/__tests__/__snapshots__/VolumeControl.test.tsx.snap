// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VolumeControls > renders with default volume 1`] = `
<div>
  <div
    aria-label="Volume controls"
    class="flex items-center gap-4"
  >
    <button
      aria-label="Mute"
      class="text-primary-dark transition-colors hover:text-primary-sage dark:text-primary-light dark:hover:text-primary-sage"
    >
      <svg
        class="lucide lucide-volume2 h-6 w-6"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
        />
        <path
          d="M16 9a5 5 0 0 1 0 6"
        />
        <path
          d="M19.364 18.364a9 9 0 0 0 0-12.728"
        />
      </svg>
    </button>
    <div
      aria-label="Volume slider"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="50"
      class="relative h-2 w-[374px] cursor-pointer rounded-lg bg-primary-light dark:bg-primary-sage/30"
      role="slider"
    >
      <div
        class="absolute h-full rounded-l-lg bg-primary-sage transition-[width] duration-75 dark:bg-primary-light"
        style="width: 50%;"
      />
      <div
        class="absolute top-1/2 h-4 w-4 -translate-y-1/2 rounded-full bg-primary-sage shadow-md transition-transform duration-75 hover:scale-110 dark:bg-primary-light"
        style="left: 50%; transform: translate(-50%, -50%);"
      />
    </div>
  </div>
</div>
`;

exports[`VolumeControls > renders with max volume 1`] = `
<div>
  <div
    aria-label="Volume controls"
    class="flex items-center gap-4"
  >
    <button
      aria-label="Mute"
      class="text-primary-dark transition-colors hover:text-primary-sage dark:text-primary-light dark:hover:text-primary-sage"
    >
      <svg
        class="lucide lucide-volume2 h-6 w-6"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
        />
        <path
          d="M16 9a5 5 0 0 1 0 6"
        />
        <path
          d="M19.364 18.364a9 9 0 0 0 0-12.728"
        />
      </svg>
    </button>
    <div
      aria-label="Volume slider"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="100"
      class="relative h-2 w-[374px] cursor-pointer rounded-lg bg-primary-light dark:bg-primary-sage/30"
      role="slider"
    >
      <div
        class="absolute h-full rounded-l-lg bg-primary-sage transition-[width] duration-75 dark:bg-primary-light"
        style="width: 100%;"
      />
      <div
        class="absolute top-1/2 h-4 w-4 -translate-y-1/2 rounded-full bg-primary-sage shadow-md transition-transform duration-75 hover:scale-110 dark:bg-primary-light"
        style="left: 100%; transform: translate(-50%, -50%);"
      />
    </div>
  </div>
</div>
`;

exports[`VolumeControls > renders with zero volume 1`] = `
<div>
  <div
    aria-label="Volume controls"
    class="flex items-center gap-4"
  >
    <button
      aria-label="Mute"
      class="text-primary-dark transition-colors hover:text-primary-sage dark:text-primary-light dark:hover:text-primary-sage"
    >
      <svg
        class="lucide lucide-volume2 h-6 w-6"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11 4.702a.705.705 0 0 0-1.203-.498L6.413 7.587A1.4 1.4 0 0 1 5.416 8H3a1 1 0 0 0-1 1v6a1 1 0 0 0 1 1h2.416a1.4 1.4 0 0 1 .997.413l3.383 3.384A.705.705 0 0 0 11 19.298z"
        />
        <path
          d="M16 9a5 5 0 0 1 0 6"
        />
        <path
          d="M19.364 18.364a9 9 0 0 0 0-12.728"
        />
      </svg>
    </button>
    <div
      aria-label="Volume slider"
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="0"
      class="relative h-2 w-[374px] cursor-pointer rounded-lg bg-primary-light dark:bg-primary-sage/30"
      role="slider"
    >
      <div
        class="absolute h-full rounded-l-lg bg-primary-sage transition-[width] duration-75 dark:bg-primary-light"
        style="width: 0%;"
      />
      <div
        class="absolute top-1/2 h-4 w-4 -translate-y-1/2 rounded-full bg-primary-sage shadow-md transition-transform duration-75 hover:scale-110 dark:bg-primary-light"
        style="left: 0%; transform: translate(-50%, -50%);"
      />
    </div>
  </div>
</div>
`;
